src/carlib/car.rs

```diff
@@ -3,12 +3,26 @@
 // accessible from the outside.
 
 // A trait is similar to an interface in Java or C#.
+pub const UNNAMED_CAR: &str = "<unnamed>";
 pub trait Car {
     // I want these methods to manipulate the speed
     // of the car based on an acceleration property.
-    fn accelerate(&mut self, duration: f64);
-    fn brake(&mut self, force: f64);
+    fn get_acceleration(&self) -> f64;
     fn get_speed(&self) -> f64;
+    fn set_speed(&mut self, new_speed: f64);
```

Comment:
The speed of a car shall not be able to be set directly. It needs to be accelerated to reach a certain speed.

Reply:
I know, its just code, but for sake of the exercise I rather stick to constraints like this.

Reply:
You can't mark functions in traits as private, but you can create private/public traits that only exposes the functionality that you want and hide the ones that you don't want.

In this case, set_speed is only used internally, but that `Car1` type in main cannot use it.
```rust
//not public
mod inner_lib {
    trait InnerCar {
        fn get_acceleration(&self) -> f64;
        fn get_speed(&self) -> f64;
        fn set_speed(&mut self, new_speed: f64);
    }
    impl<T> crate::Car for T
    where
        T: InnerCar,
    {
        fn get_acceleration(&self) -> f64 {
            <Self as InnerCar>::get_acceleration(self)
        }
        fn get_speed(&self) -> f64 {
            <Self as InnerCar>::get_speed(self)
        }
        fn accelerate(&mut self, duration: f64) {
            let speed = self.get_speed() + (self.get_acceleration() * duration);
            self.set_speed(speed);
        }
    }
    impl InnerCar for crate::Car1 {
        fn get_acceleration(&self) -> f64 {
            10.0
        }
        fn get_speed(&self) -> f64 {
            self.0
        }
        fn set_speed(&mut self, new_speed: f64) {
            self.0 = new_speed;
        }
    }
}
//trait that is exposed to all classes
trait Car {
    fn get_acceleration(&self) -> f64;
    fn get_speed(&self) -> f64;
    fn accelerate(&mut self, duration: f64);
}

pub struct Car1(f64);

fn main() {
    let mut car = Car1(10.2);
    let car_dyn: &mut dyn Car = &mut car;
    car_dyn.accelerate(10.0);
    //car_dyn.set_speed(11.0); //ERROR, set_speed is in the private trait
    println!("car speed {}", car_dyn.get_speed());
}

```

Reply:
BTW I think this is 50% a great idea and 50% a huge mistake. :smile: